features/flows/elastic/src/main/java/org/opennms/netmgt/flows/elastic/thresholding/FlowThresholding.java

```diff
@@ -0,0 +1,135 @@
+        this.thresholdingSessions = new CacheBuilder<NodeInterfaceKey, ThresholdingSession>()
+                .withConfig(sessionCacheConfig)
+                .withCacheLoader(new CacheLoader<NodeInterfaceKey, ThresholdingSession>() {
+                    @Override
+                    public ThresholdingSession load(final NodeInterfaceKey nik) throws Exception {
+                        return FlowThresholding.this.thresholdingService.createSession(nik.nodeId,
+                                                                                       nik.ifaceAddr,
+                                                                                       SERVICE_NAME,
+                                                                                       FLOW_APP_RRD_REPO,
+                                                                                       new ServiceParameters(Collections.emptyMap()));
+                    }
+                }).build();
+    }
+
+    public void threshold(final List<FlowDocument> documents,
+                          final FlowSource source) throws ExecutionException, ThresholdInitializationException {
+        for (final var document : documents) {
+            if (document.getNodeExporter() != null && !Strings.isNullOrEmpty(document.getApplication())) {
+                final var nodeId = document.getNodeExporter().getNodeId();
+
+                final var key = new ApplicationKey(nodeId,
+                                                   document.getDirection() == Direction.INGRESS
+                                                   ? document.getInputSnmp()
+                                                   : document.getOutputSnmp(),
+                                                   document.getApplication());
+
+                // Update the counter
+                final var counter = this.applicationAccumulator.computeIfAbsent(key, k -> new AtomicLong(0))
+                                                               .addAndGet(document.getBytes());
+
+                // Apply thresholding
+                final var thresholdingSession = this.thresholdingSessions.get(new NodeInterfaceKey(nodeId, source.getSourceAddress()));
+
+                final var collectionAgent = this.collectionAgentFactory.createCollectionAgent(
+                        Integer.toString(nodeId),
+                        InetAddressUtils.addr(source.getSourceAddress()));
+
+                final var nodeResource = new NodeLevelResource(nodeId);
+                final var appResource = new DeferredGenericTypeResource(nodeResource, RESOURCE_TYPE_NAME, document.getApplication());
+
+                final var collectionSetBuilder = new CollectionSetBuilder(collectionAgent)
+                        .withCounter(appResource, "flowappxxxx", "bytes", counter);
+                // TODO fooker: Set sequence number from flow to aid distributed thresholding
+
+                thresholdingSession.accept(collectionSetBuilder.build());
```

Comment:
Creating a `collectionAgent`, `nodeResource`, and `appResource` each time anew seems expensive to me. Could these things be cached too? The `threasholdingSession` map could contain ready to call `consumers` that setup all these things when a new entry is added to the map.

Reply:
Refactored to cache the `CollectionAgent` along with the `ThresholdingSession`. The resources are to cheap to create to bother right now. They are even subject of change in further development.